cmd/bucket-replication.go

```diff
@@ -1349,53 +1349,6 @@ func (ri ReplicateObjectInfo) replicateAll(ctx context.Context, objectAPI Object
 	}()
 
 	rAction = replicateAll
-	oi, cerr := tgt.StatObject(ctx, tgt.Bucket, object, minio.StatObjectOptions{
-		VersionID: objInfo.VersionID,
-		Internal: minio.AdvancedGetOptions{
-			ReplicationProxyRequest: "false",
-		},
-	})
-	if cerr == nil {
-		rAction = getReplicationAction(objInfo, oi, ri.OpType)
-		rinfo.ReplicationStatus = replication.Completed
-		if rAction == replicateNone {
-			if ri.OpType == replication.ExistingObjectReplicationType &&
-				objInfo.ModTime.Unix() > oi.LastModified.Unix() && objInfo.VersionID == nullVersionID {
-				logger.LogIf(ctx, fmt.Errorf("unable to replicate %s/%s (null). Newer version exists on target", bucket, object))
-				sendEvent(eventArgs{
-					EventName:  event.ObjectReplicationNotTracked,
-					BucketName: bucket,
-					Object:     objInfo,
-					UserAgent:  "Internal: [Replication]",
-					Host:       globalLocalNodeName,
-				})
-			}
-			// object with same VersionID already exists, replication kicked off by
-			// PutObject might have completed
-			if objInfo.TargetReplicationStatus(tgt.ARN) == replication.Pending || objInfo.TargetReplicationStatus(tgt.ARN) == replication.Failed || ri.OpType == replication.ExistingObjectReplicationType {
-				// if metadata is not updated for some reason after replication, such as
-				// 503 encountered while updating metadata - make sure to set ReplicationStatus
-				// as Completed.
-				//
-				// Note: Replication Stats would have been updated despite metadata update failure.
-				rinfo.ReplicationAction = rAction
-				rinfo.ReplicationStatus = replication.Completed
-			}
-			return
-		}
-	}
-	// if target returns error other than NoSuchKey, defer replication attempt
-	if cerr != nil && minio.ToErrorResponse(cerr).Code != "NoSuchKey" && minio.ToErrorResponse(cerr).Code != "NoSuchVersion" {
-		logger.LogIf(ctx, fmt.Errorf("unable to replicate %s/%s (%s). Target returned %s error on HEAD", bucket, object, objInfo.VersionID, cerr))
-		sendEvent(eventArgs{
-			EventName:  event.ObjectReplicationNotTracked,
-			BucketName: bucket,
-			Object:     objInfo,
-			UserAgent:  "Internal: [Replication]",
-			Host:       globalLocalNodeName,
-		})
-		return
-	}
```

Comment:
if the HEAD is removed, every metadata update is a full overwrite - this is not ok. We already optimize to do a HEAD only for replication failure healing, not for inline replication